scarb/src/core/checksum.rs

```diff
@@ -0,0 +1,171 @@
+use sha2::Digest as _;
+
+#[derive(Clone, PartialEq, Eq, Hash, Serialize, Deserialize)]
+#[serde(try_from = "String", into = "String")]
+pub struct Checksum([u8; 32]);
+
+impl Checksum {
+    const HASH_FUNC_TYPE: &'_ str = "sha256";
+    const ENCODING: Encoding = HEXLOWER_PERMISSIVE;
+
+    pub fn parse(s: &str) -> Result<Self> {
+        fn inner(s: &str) -> Result<Checksum> {
+            let Some((hash_func_type, hash)) = s.split_once(':') else {
+                bail!("checksum is missing hash function type prefix");
+            };
+
+            ensure!(
+                hash_func_type == Checksum::HASH_FUNC_TYPE,
+                "unsupported hash function type: {hash_func_type}",
+            );
+
+            let mut buffer = [0u8; 32];
+            let expected_len = buffer.len();
+
+            let decode_len = Checksum::ENCODING.decode_len(hash.len())?;
+            ensure!(
+                decode_len == expected_len,
+                "invalid checksum length {decode_len}, should be {expected_len}"
+            );
+
+            let len = Checksum::ENCODING
+                .decode_mut(hash.as_bytes(), &mut buffer)
+                .map_err(|e| e.error)?;
+            ensure!(
+                len == expected_len,
+                "invalid checksum length {len}, should be {expected_len}"
+            );
+
+            Ok(Checksum(buffer))
+        }
+
+        inner(s).with_context(|| format!("failed to parse checksum: {s}"))
+    }
+
+    /// Create a [`Digest`] instance which will use the same algorithm as this checksum.
+    pub fn digest(&self) -> Digest {
+        Digest::recommended()
+    }
+}
+
+impl FromStr for Checksum {
+    type Err = anyhow::Error;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        Checksum::parse(s)
+    }
+}
+
+impl<'a> TryFrom<&'a str> for Checksum {
+    type Error = anyhow::Error;
+
+    fn try_from(s: &str) -> Result<Self, Self::Error> {
+        s.parse()
+    }
+}
+
+impl TryFrom<String> for Checksum {
+    type Error = anyhow::Error;
+
+    fn try_from(s: String) -> Result<Self, Self::Error> {
+        s.parse()
+    }
+}
+
+impl From<Checksum> for String {
+    fn from(c: Checksum) -> Self {
+        c.to_string()
+    }
+}
+
+impl fmt::Display for Checksum {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.write_str(Checksum::HASH_FUNC_TYPE)?;
+        f.write_char(':')?;
+
+        let mut buffer = [0u8; 64];
+        Checksum::ENCODING.encode_mut(&self.0, &mut buffer);
+        // SAFETY: We just generated this hexadecimal string.
+        let string = unsafe { str::from_utf8_unchecked(&buffer) };
+        f.write_str(string)?;
+
+        Ok(())
+    }
+}
+
+impl fmt::Debug for Checksum {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "Checksum({self})")
+    }
+}
+
+pub struct Digest(sha2::Sha256);
+
+impl Digest {
+    /// Get recommended digest algorithm.
+    pub fn recommended() -> Self {
```

Comment:
Why not just default? 🤔

Reply:
I have thought about this and I concluded `default` is not precise enough? It creates an incentive to use default everywhere, while programmers should carefully decide whether they want to create a digest using the currently preferred algorithm, or create one with an algorithm matching the checksum we got to verify.